pkg/skaffold/schema/versions_test.go

```diff
@@ -710,8 +715,111 @@ func TestParseConfigAndUpgradeToOlderVersion(t *testing.T) {
+		expected    string
+		shouldErr   bool
+		err         error
+	}{
+		{
+			apiVersions: []util.VersionedConfig{
+				&v1alpha1.SkaffoldConfig{APIVersion: v1alpha1.Version},
+				&v1beta1.SkaffoldConfig{APIVersion: v1beta1.Version},
+				&v2alpha1.SkaffoldConfig{APIVersion: v2alpha1.Version},
+				&v2beta1.SkaffoldConfig{APIVersion: v2beta1.Version},
+			},
+			description: "valid compatibility check for all v1 schemas releases",
+			expected:    latest_v1.Version,
+			shouldErr:   false,
+		},
+		{
+
+			apiVersions: []util.VersionedConfig{
+				&v3alpha1.SkaffoldConfig{APIVersion: v3alpha1.Version},
+			},
+			description: "valid compatibility check for all v2 schemas releases",
+			expected:    latest_v2.Version,
+			shouldErr:   false,
+		},
+		{
+			apiVersions: []util.VersionedConfig{
+				&v1alpha1.SkaffoldConfig{APIVersion: v1alpha1.Version},
+				&v3alpha1.SkaffoldConfig{APIVersion: v3alpha1.Version},
+			},
+			description: "invalid compatibility among v1 and v2 versions",
+			shouldErr:   true,
+			err: fmt.Errorf("detected incompatible versions:%v are incompatible with %v",
+				[]string{latest_v1.Version, v1alpha1.Version}, []string{v3alpha1.Version}),
+		},
+		{
+			apiVersions: []util.VersionedConfig{
+				&v1alpha1.SkaffoldConfig{APIVersion: "vXalphaY"},
+			},
+			description: "invalid api version",
+			shouldErr:   true,
+			err:         fmt.Errorf("unknown apiVersion vXalpaY"),
+		},
+	}
+	for _, test := range tests {
+		testutil.Run(t, test.description, func(t *testutil.T) {
+			upToDateVersion, err := checkCompatibility(test.apiVersions)
+			t.CheckErrorAndDeepEqual(test.shouldErr, err, test.expected, upToDateVersion)
+		})
+	}
+}
+
+func TestIsCompatibleWith(t *testing.T) {
+	tests := []struct {
+		description string
+		apiVersions []util.VersionedConfig
+		toVersion   string
+		shouldErr   bool
+		err         error
+	}{
+		{
+			apiVersions: []util.VersionedConfig{
+				&v1alpha1.SkaffoldConfig{APIVersion: v1alpha1.Version},
+				&v1beta1.SkaffoldConfig{APIVersion: v1beta1.Version},
+				&v2alpha1.SkaffoldConfig{APIVersion: v2alpha1.Version},
+				&v2beta1.SkaffoldConfig{APIVersion: v2beta1.Version},
+			},
+			description: "v1 schemas are compatible to a v1 schema",
+			toVersion:   v2beta14.Version,
+			shouldErr:   false,
+		},
+		{
+			apiVersions: []util.VersionedConfig{
+				&v3alpha1.SkaffoldConfig{APIVersion: v3alpha1.Version},
+			},
+			description: "v2 schemas are compatible to a v2 schema",
+			toVersion:   latest_v2.Version,
+			shouldErr:   false,
+		},
+		{
+			apiVersions: []util.VersionedConfig{
+				&v1alpha1.SkaffoldConfig{APIVersion: v1alpha1.Version},
+				&v1beta1.SkaffoldConfig{APIVersion: v1beta1.Version},
+			},
+			description: "v1 and v2 schemas are incompatible.",
+			toVersion:   latest_v2.Version,
+			shouldErr:   true,
+			err: fmt.Errorf("the following versions are incompatible with target version %v. upgrade aborted",
+				[]string{v1alpha1.Version, v1beta1.Version}),
+		},
```

Comment:
lmk if you think it's unnecessary, but might be good to test v2 -> v1 incompatible upgrade as well

Reply:
Done.